### ISSUE TITLE
`and` clause in filter types is order-sensitive

### ISSUE THREAD

**Issue by jsmaniac** [racket/typed-racket#330]
### What version of Racket are you using?
6.4.0.15 (20160407-4bcb657)

### What program did you run?
```
#lang typed/racket
(: fn (∀ (A B) (→ (U (Pairof B (Listof A))
                     (Pairof A (Listof (U A B))))
                  (→ (U A B) Any : #:+ (and B (! A))
                     #:- (! B))
                  Any)))
(define (fn l pred?)
  (if (pred? (car l))
      (ann l (Pairof B (Listof A))) ;; ERROR HERE
      (void)))
```

The program above fails to typecheck, because `l` is inferred to have the type `(Pairof B (Listof (U A B)))` (which can never happen).

Swapping the filters in `(and B (! A))` on line 4 makes the program typecheck:

```
#lang typed/racket
(: fn (∀ (A B) (→ (U (Pairof B (Listof A))
                     (Pairof A (Listof (U A B))))
                  (→ (U A B) Any : #:+ (and (! A) B) ;; Swapped
                     #:- (! B))
                  Any)))
(define (fn l pred?)
  (if (pred? (car l))
      (ann l (Pairof B (Listof A))) ;; Okay
      (void)))
```

### What should have happened?

Both programs should typecheck. `(and …)` filter clauses shouldn't be sensitive to the order of the sub-filters.

### If you got an error message, please include it here.

```
 Type Checker: type mismatch
  expected: (Pairof B (Listof A))
  given: (Pairof B (Listof (U A B))) in: l
```

**Comment by jsmaniac** [comment]
Also, the following program does not typecheck:

```
(: fn (∀ (A B) (→ (→ Any Any : #:+ (and (! A) B))
                  Any)))
(define (fn pred?)
  (fn pred?))
```

It gives this (rather confusing) error. Removing either filter inside the `(and (! A) B)` makes it typecheck.
```
Type Checker: Polymorphic function `fn' could not be applied to arguments:
Argument 1:
  Expected: (-> Any Any)
  Given:    (-> Any Any)

Result type:     Any
Expected result: Any
 in: (fn pred?)
```

**Comment by samth** [comment]
I think fixing this requires a fairly comprehensive rewrite of the proof engine in TR, which @andmkent has planned for the future.

**Comment by jsmaniac** [comment]
Interestingly, for the recursive `fn` in my comment, using `let rec` makes it typecheck. Do these issues have both the same source, or should I file the recursive-call as a separate one?

```
#lang typed/racket
(: fn (∀ (A B) (→ (→ Any Any : #:+ (and (! A) B))
                  Any)))
(define (fn pred?)
  (let rec [(pred? pred?)]
    (rec pred?)))                ;; Works
```

**Comment by samth** [comment]
I'm not sure if they're the same issue.

**Comment by jsmaniac** [comment]
I moved the comments above to a separate issue (#333), feel free to close either if they have the same cause in the end.

**Comment by andmkent** [comment]
At the moment I'm exploring how intersection types might fit in to Typed Racket ( #350 )--- it looks like there addition allows TR to type check this program regardless of the order of ```(!A)``` and ```B``` in the conjunction.

**Comment by jsmaniac** [comment]
The original issue seems fixed in the latest version (6.6.0.5), so I'm closing this. #333 is still failing however.

Thank you @andmkent for the fix!
